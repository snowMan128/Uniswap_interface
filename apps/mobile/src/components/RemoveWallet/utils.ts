import { CommonActions } from '@react-navigation/core'
import { dispatchNavigationAction } from 'src/app/navigation/rootNavigation'
import { OnboardingScreens, Screens } from 'src/screens/Screens'
import { ImportType, OnboardingEntryPoint } from 'wallet/src/features/onboarding/types'

// This fast-forwards user to the same app state as if
// they have pressed "Get Started" on Landing and should now see import method view
export function navigateToOnboardingImportMethod(): void {
  dispatchNavigationAction(
    CommonActions.reset({
      index: 0,
      routes: [
        {
          name: Screens.OnboardingStack,
          state: {
            index: 1,
            routes: [
              {
                name: OnboardingScreens.Landing,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
              {
                name: OnboardingScreens.ImportMethod,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
            ],
          },
        },
      ],
    })
  )
}
