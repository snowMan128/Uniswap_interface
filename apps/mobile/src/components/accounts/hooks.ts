import { NetworkStatus, WatchQueryFetchPolicy } from '@apollo/client'
import {
  AccountListQuery,
  // eslint-disable-next-line no-restricted-imports
  useAccountListQuery,
} from 'wallet/src/data/__generated__/types-and-hooks'
// eslint-disable-next-line no-restricted-imports
import { usePortfolioValueModifiers } from 'wallet/src/features/dataApi/balances'
import { GqlResult } from 'wallet/src/features/dataApi/types'

export function useAccountList({
  addresses,
  fetchPolicy,
  notifyOnNetworkStatusChange,
}: {
  addresses: Address | Address[]
  fetchPolicy?: WatchQueryFetchPolicy
  notifyOnNetworkStatusChange?: boolean | undefined
}): GqlResult<AccountListQuery> & {
  startPolling: (pollInterval: number) => void
  stopPolling: () => void
  networkStatus: NetworkStatus
  refetch: () => void
} {
  const valueModifiers = usePortfolioValueModifiers(addresses)
  const { data, loading, networkStatus, refetch, startPolling, stopPolling } = useAccountListQuery({
    variables: { addresses, valueModifiers },
    notifyOnNetworkStatusChange,
    fetchPolicy,
  })

  return {
    data,
    loading,
    networkStatus,
    refetch,
    startPolling,
    stopPolling,
  }
}
