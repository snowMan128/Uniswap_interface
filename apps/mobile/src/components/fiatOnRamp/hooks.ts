import { useIsDarkMode } from 'ui/src'
import { FORLogo } from 'wallet/src/features/fiatOnRamp/types'

export function useFiatOnRampLogoUrl(logos: FORLogo | undefined): string | undefined {
  const isDarkMode = useIsDarkMode()

  if (!logos) {
    return
  }

  return isDarkMode ? logos.darkLogo : logos.lightLogo
}
