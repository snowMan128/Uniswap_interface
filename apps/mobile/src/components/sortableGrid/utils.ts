import { FlatList, ScrollView } from 'react-native'

const hasProp = <O extends object, P extends string>(
  object: O,
  prop: P
): object is O & Record<P, unknown> => {
  return prop in object
}

export const defaultKeyExtractor = <I>(item: I, index: number): string => {
  if (typeof item === 'string') {
    return item
  }

  if (typeof item === 'object' && item !== null) {
    if (hasProp(item, 'id')) {
      return String(item.id)
    }
    if (hasProp(item, 'key')) {
      return String(item.key)
    }
  }

  return String(index)
}

export const isScrollView = (scrollable: ScrollView | FlatList): scrollable is ScrollView => {
  return 'scrollTo' in scrollable
}
