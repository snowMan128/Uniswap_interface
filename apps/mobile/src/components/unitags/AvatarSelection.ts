import { ImageLibraryOptions, launchImageLibrary } from 'react-native-image-picker'
import { NUM_FIRST_NFTS } from 'wallet/src/components/nfts/NftsList'
import { useNftsTabQuery } from 'wallet/src/data/__generated__/types-and-hooks'
import { formatNftItems } from 'wallet/src/features/nfts/utils'

// Selected image will be shrunk to max width/height
// URI will then be for an image of those dimensions
const IMAGE_OPTIONS: ImageLibraryOptions = {
  mediaType: 'photo',
  maxWidth: 500,
  maxHeight: 500,
  quality: 1, // best quality
  includeBase64: false,
  selectionLimit: 1,
}

export async function selectPhotoFromLibrary(): Promise<string | undefined> {
  const response = await launchImageLibrary(IMAGE_OPTIONS)
  if (!response.didCancel && !response.errorCode && response.assets) {
    return response.assets[0]?.uri
  }
}

export function useAvatarSelectionHandler({
  address,
  avatarImageUri,
  setAvatarImageUri,
  showModal,
}: {
  address: string
  avatarImageUri: string | undefined
  setAvatarImageUri: (uri: string) => void
  showModal: () => void
}): { avatarSelectionHandler: () => Promise<void>; hasNFTs: boolean } {
  const { data: nftsData } = useNftsTabQuery({
    variables: { ownerAddress: address, first: NUM_FIRST_NFTS, filter: { filterSpam: false } },
  })
  const nftItems = formatNftItems(nftsData)

  const hasNFTs = nftItems !== undefined && nftItems?.length > 0
  const hasAvatarImage = avatarImageUri && avatarImageUri !== ''

  if (hasNFTs || hasAvatarImage) {
    return { avatarSelectionHandler: async () => showModal(), hasNFTs }
  } else {
    return {
      avatarSelectionHandler: async (): Promise<void> => {
        const selectedPhoto = await selectPhotoFromLibrary()
        if (selectedPhoto) {
          setAvatarImageUri(selectedPhoto)
        }
      },
      hasNFTs,
    }
  }
}
