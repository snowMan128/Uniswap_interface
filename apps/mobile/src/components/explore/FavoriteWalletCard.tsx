import { ImpactFeedbackStyle } from 'expo-haptics'
import { default as React, useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { ViewProps } from 'react-native'
import ContextMenu from 'react-native-context-menu-view'
import { useAppDispatch } from 'src/app/hooks'
import { useEagerExternalProfileNavigation } from 'src/app/navigation/hooks'
import RemoveButton from 'src/components/explore/RemoveButton'
import { disableOnPress } from 'src/utils/disableOnPress'
import { Flex, TouchableArea } from 'ui/src'
import { borderRadii, iconSizes } from 'ui/src/theme'
import { AccountIcon } from 'wallet/src/components/accounts/AccountIcon'
import { DisplayNameText } from 'wallet/src/components/accounts/DisplayNameText'
import { BaseCard } from 'wallet/src/components/BaseCard/BaseCard'
import { removeWatchedAddress } from 'wallet/src/features/favorites/slice'
import { useAvatar, useDisplayName } from 'wallet/src/features/wallet/hooks'
import { DisplayNameType } from 'wallet/src/features/wallet/types'

type FavoriteWalletCardProps = {
  address: Address
  isEditing?: boolean
  setIsEditing: (update: boolean) => void
} & ViewProps

export default function FavoriteWalletCard({
  address,
  isEditing,
  setIsEditing,
  ...rest
}: FavoriteWalletCardProps): JSX.Element {
  const { t } = useTranslation()
  const dispatch = useAppDispatch()
  const { preload, navigate } = useEagerExternalProfileNavigation()

  const displayName = useDisplayName(address)
  const { avatar } = useAvatar(address)

  const icon = useMemo(() => {
    return <AccountIcon address={address} avatarUri={avatar} size={iconSizes.icon20} />
  }, [address, avatar])

  const onRemove = useCallback(() => {
    dispatch(removeWatchedAddress({ address }))
  }, [address, dispatch])

  /// Options for long press context menu
  const menuActions = useMemo(() => {
    return [
      { title: t('Remove favorite'), systemIcon: 'heart.fill' },
      { title: t('Edit favorites'), systemIcon: 'square.and.pencil' },
    ]
  }, [t])

  return (
    <ContextMenu
      actions={menuActions}
      disabled={isEditing}
      style={{ borderRadius: borderRadii.rounded16 }}
      onPress={(e): void => {
        // Emitted index based on order of menu action array
        // remove favorite action
        if (e.nativeEvent.index === 0) {
          onRemove()
        }
        // Edit mode toggle action
        if (e.nativeEvent.index === 1) {
          setIsEditing(true)
        }
      }}
      {...rest}>
      <TouchableArea
        hapticFeedback
        borderRadius="$rounded16"
        hapticStyle={ImpactFeedbackStyle.Light}
        m="$spacing4"
        onLongPress={disableOnPress}
        onPress={(): void => {
          navigate(address)
        }}
        onPressIn={async (): Promise<void> => {
          await preload(address)
        }}>
        <BaseCard.Shadow>
          <Flex row gap="$spacing4" justifyContent="space-between">
            <Flex row shrink alignItems="center" gap="$spacing8">
              {icon}
              <DisplayNameText
                displayName={displayName}
                textProps={{
                  adjustsFontSizeToFit: displayName?.type === DisplayNameType.Address,
                  variant: 'body1',
                }}
              />
            </Flex>
            <RemoveButton visible={isEditing} onPress={onRemove} />
          </Flex>
        </BaseCard.Shadow>
      </TouchableArea>
    </ContextMenu>
  )
}
