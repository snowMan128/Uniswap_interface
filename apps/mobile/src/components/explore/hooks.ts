import { SharedEventName } from '@uniswap/analytics-events'
import { useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { NativeSyntheticEvent, Share } from 'react-native'
import { ContextMenuAction, ContextMenuOnPressNativeEvent } from 'react-native-context-menu-view'
import { useSelectHasTokenFavorited, useToggleFavoriteCallback } from 'src/features/favorites/hooks'
import { openModal } from 'src/features/modals/modalSlice'
import { sendMobileAnalyticsEvent } from 'src/features/telemetry'
import { MobileEventName, ShareableEntity } from 'src/features/telemetry/constants'
import { logger } from 'utilities/src/logger/logger'
import { ChainId } from 'wallet/src/constants/chains'
import { AssetType } from 'wallet/src/entities/assets'
import { useCopyTokenAddressCallback } from 'wallet/src/features/tokens/hooks'
import {
  CurrencyField,
  TransactionState,
} from 'wallet/src/features/transactions/transactionState/types'
import { useAppDispatch } from 'wallet/src/state'
import { ElementName, ModalName, SectionNameType } from 'wallet/src/telemetry/constants'
import { getTokenUrl } from 'wallet/src/utils/linking'

import { CurrencyId, currencyIdToAddress } from 'wallet/src/utils/currencyId'

interface TokenMenuParams {
  currencyId: CurrencyId
  chainId: ChainId
  analyticsSection: SectionNameType
  // token, which are in favorite section would have it defined
  onEditFavorites?: () => void
}

// Provide context menu related data for token
export function useExploreTokenContextMenu({
  currencyId,
  chainId,
  analyticsSection,
  onEditFavorites,
}: TokenMenuParams): {
  menuActions: Array<ContextMenuAction & { onPress: () => void }>
  onContextMenuPress: (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>) => void
} {
  const { t } = useTranslation()
  const isFavorited = useSelectHasTokenFavorited(currencyId)
  const dispatch = useAppDispatch()

  // `address` is undefined for native currencies, so we want to extract it from
  // currencyId, where we have hardcoded addresses for native currencies
  const currencyAddress = currencyIdToAddress(currencyId)

  const onPressCopyContractAddress = useCopyTokenAddressCallback(currencyAddress)

  const onPressShare = useCallback(async () => {
    const tokenUrl = getTokenUrl(currencyId)
    if (!tokenUrl) {
      return
    }
    try {
      await Share.share({
        message: tokenUrl,
      })
      sendMobileAnalyticsEvent(MobileEventName.ShareButtonClicked, {
        entity: ShareableEntity.Token,
        url: tokenUrl,
      })
    } catch (error) {
      logger.error(error, { tags: { file: 'balances/hooks.ts', function: 'onPressShare' } })
    }
  }, [currencyId])

  const toggleFavoriteToken = useToggleFavoriteCallback(currencyId, isFavorited)

  const onPressSwap = useCallback(() => {
    const swapFormState: TransactionState = {
      exactCurrencyField: CurrencyField.INPUT,
      exactAmountToken: '0',
      [CurrencyField.INPUT]: null,
      [CurrencyField.OUTPUT]: {
        chainId,
        address: currencyAddress,
        type: AssetType.Currency,
      },
    }
    dispatch(openModal({ name: ModalName.Swap, initialState: swapFormState }))
    sendMobileAnalyticsEvent(SharedEventName.ELEMENT_CLICKED, {
      element: ElementName.Swap,
      section: analyticsSection,
    })
  }, [analyticsSection, chainId, currencyAddress, dispatch])

  const onPressToggleFavorite = useCallback(() => {
    toggleFavoriteToken()
  }, [toggleFavoriteToken])

  const menuActions = useMemo(
    () => [
      {
        title: isFavorited ? t('Remove favorite') : t('Favorite token'),
        systemIcon: isFavorited ? 'heart.fill' : 'heart',
        onPress: onPressToggleFavorite,
      },
      ...(onEditFavorites
        ? [
            {
              title: t('Edit favorites'),
              systemIcon: 'square.and.pencil',
              onPress: onEditFavorites,
            },
          ]
        : []),
      { title: t('Swap'), systemIcon: 'arrow.2.squarepath', onPress: onPressSwap },
      {
        title: t('Copy contract address'),
        systemIcon: 'doc.on.doc',
        onPress: onPressCopyContractAddress,
      },
      ...(!onEditFavorites
        ? [
            {
              title: t('Share'),
              systemIcon: 'square.and.arrow.up',
              onPress: onPressShare,
            },
          ]
        : []),
    ],
    [
      isFavorited,
      t,
      onPressToggleFavorite,
      onEditFavorites,
      onPressSwap,
      onPressCopyContractAddress,
      onPressShare,
    ]
  )

  const onContextMenuPress = useCallback(
    async (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>): Promise<void> => {
      await menuActions[e.nativeEvent.index]?.onPress?.()
    },
    [menuActions]
  )

  return { menuActions, onContextMenuPress }
}
