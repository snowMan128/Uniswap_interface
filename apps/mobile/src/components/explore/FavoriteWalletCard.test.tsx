import { makeMutable } from 'react-native-reanimated'
import configureMockStore from 'redux-mock-store'
import { preloadedMobileState } from 'src/test/fixtures'
import { fireEvent, render, waitFor } from 'src/test/test-utils'
import * as unitagHooks from 'uniswap/src/features/unitags/hooks'
import { MobileScreens } from 'uniswap/src/types/screens/mobile'
import * as ensHooks from 'wallet/src/features/ens/api'
import {
  ON_PRESS_EVENT_PAYLOAD,
  SAMPLE_SEED_ADDRESS_1,
  preloadedWalletState,
  signerMnemonicAccount,
} from 'wallet/src/test/fixtures'
import { sanitizeAddressText, shortenAddress } from 'wallet/src/utils/addresses'
import FavoriteWalletCard, { FavoriteWalletCardProps } from './FavoriteWalletCard'

const mockedNavigation = {
  navigate: jest.fn(),
}

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native')
  return {
    ...actualNav,
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    useNavigation: () => mockedNavigation,
  }
})

const mockStore = configureMockStore()

const defaultProps: FavoriteWalletCardProps = {
  address: SAMPLE_SEED_ADDRESS_1,
  pressProgress: makeMutable(0),
  isEditing: false,
  dragActivationProgress: makeMutable(0),
  setIsEditing: jest.fn(),
}

describe('FavoriteWalletCard', () => {
  it('renders without error', () => {
    const tree = render(<FavoriteWalletCard {...defaultProps} />)

    expect(tree).toMatchSnapshot()
  })

  describe('displayName', () => {
    afterEach(() => {
      jest.restoreAllMocks()
    })

    it('renders unitag name if available', () => {
      jest.spyOn(unitagHooks, 'useUnitagByAddress').mockReturnValue({
        unitag: { username: 'unitagname' },
        loading: false,
      })

      const { queryByText } = render(<FavoriteWalletCard {...defaultProps} />)

      expect(queryByText('unitagname')).toBeTruthy()
    })

    it('renders ens name if available', () => {
      jest.spyOn(ensHooks, 'useENSName').mockReturnValue({
        data: 'ensname.eth',
        loading: false,
        error: undefined,
      })

      const { queryByText } = render(<FavoriteWalletCard {...defaultProps} />)

      expect(queryByText('ensname.eth')).toBeTruthy()
    })

    it('renders local name if wallet name is set locally', () => {
      const { queryByText } = render(<FavoriteWalletCard {...defaultProps} />, {
        preloadedState: preloadedMobileState({
          wallet: preloadedWalletState({
            account: signerMnemonicAccount({
              address: defaultProps.address,
              name: 'Local account',
            }),
          }),
        }),
      })

      expect(queryByText('Local account')).toBeTruthy()
    })

    it('renders wallet address in other cases', () => {
      const { queryByText } = render(<FavoriteWalletCard {...defaultProps} />)

      const displayedAddress = sanitizeAddressText(shortenAddress(defaultProps.address))!

      expect(queryByText(displayedAddress)).toBeTruthy()
    })
  })

  describe('when not editing', () => {
    it('navigates to the wallet details screen when pressed', () => {
      const { getByTestId } = render(<FavoriteWalletCard {...defaultProps} />)

      const touchable = getByTestId('favorite-wallet-card')
      fireEvent.press(touchable, ON_PRESS_EVENT_PAYLOAD)

      expect(mockedNavigation.navigate).toHaveBeenCalledWith(MobileScreens.ExternalProfile, {
        address: defaultProps.address,
      })
    })

    it('does not display the remove button', async () => {
      const { getByTestId } = render(<FavoriteWalletCard {...defaultProps} />)

      const removeButton = getByTestId('explore/remove-button')

      await waitFor(() => {
        expect(removeButton).toHaveAnimatedStyle({ opacity: 0 })
      })
    })
  })

  describe('when editing', () => {
    it('displays the remove button', async () => {
      const { getByTestId } = render(<FavoriteWalletCard {...defaultProps} isEditing />)

      const removeButton = getByTestId('explore/remove-button')

      await waitFor(() => {
        expect(removeButton).toHaveAnimatedStyle({ opacity: 1 })
      })
    })

    it('dispatches removeWatchedAddress when remove button is pressed', () => {
      const store = mockStore({
        favorites: { tokens: [] },
        wallet: {
          accounts: {
            [defaultProps.address]: signerMnemonicAccount({ address: defaultProps.address }),
          },
        },
      })
      const { getByTestId } = render(<FavoriteWalletCard {...defaultProps} isEditing />, {
        store,
      })

      const removeButton = getByTestId('explore/remove-button')
      fireEvent.press(removeButton, ON_PRESS_EVENT_PAYLOAD)

      expect(store.getActions()).toEqual([
        {
          type: 'favorites/removeWatchedAddress',
          payload: { address: defaultProps.address },
        },
      ])
    })
  })
})
