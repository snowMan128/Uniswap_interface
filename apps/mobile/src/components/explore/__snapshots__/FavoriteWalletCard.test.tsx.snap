// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteWalletCard renders without error 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "opacity": 1,
      },
    }
  }
  collapsable={false}
  style={
    {
      "flexDirection": "column",
      "opacity": 1,
    }
  }
>
  <View
    actions={
      [
        {
          "systemIcon": "heart.fill",
          "title": "Remove favorite",
        },
        {
          "systemIcon": "square.and.pencil",
          "title": "Edit favorites",
        },
      ]
    }
    disabled={false}
    onPress={[Function]}
    style={
      {
        "borderRadius": 16,
      }
    }
  >
    <View
      cancelable={true}
      disabled={false}
      focusable={true}
      hitSlop={[Function]}
      minPressDuration={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onLongPress={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "backgroundColor": "#F9F9F9",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "flexDirection": "column",
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
          "opacity": 1,
          "transform": [
            {
              "scale": 1,
            },
          ],
        }
      }
      testID="favorite-wallet-card"
    >
      <View
        style={
          {
            "backgroundColor": "#FFFFFFff",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flexDirection": "column",
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
            "shadowColor": "rgb(0,0,0)",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.0075,
            "shadowRadius": 10,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "gap": 4,
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "flexShrink": 1,
                "gap": 8,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "transparent",
                  "borderBottomColor": "transparent",
                  "borderBottomLeftRadius": 999999,
                  "borderBottomRightRadius": 999999,
                  "borderBottomWidth": 0,
                  "borderLeftColor": "transparent",
                  "borderLeftWidth": 0,
                  "borderRightColor": "transparent",
                  "borderRightWidth": 0,
                  "borderStyle": "solid",
                  "borderTopColor": "transparent",
                  "borderTopLeftRadius": 999999,
                  "borderTopRightRadius": 999999,
                  "borderTopWidth": 0,
                  "flexDirection": "column",
                  "position": "relative",
                }
              }
              testID="account-icon"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "borderBottomLeftRadius": 999999,
                    "borderBottomRightRadius": 999999,
                    "borderTopLeftRadius": 999999,
                    "borderTopRightRadius": 999999,
                    "flexDirection": "column",
                    "height": 20,
                    "justifyContent": "center",
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": 20,
                  }
                }
              >
                <View
                  style={
                    {
                      "flexDirection": "column",
                      "height": 20,
                      "width": 20,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "flexDirection": "column",
                        "height": 20,
                        "width": 20,
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        clip={true}
                        mask={
                          <UniconMask
                            attributeData={
                              {
                                "0": "#6100FF",
                                "1": "#76D191",
                                "2": [
                                  {
                                    "fillType": "evenOdd",
                                    "path": "M31 33V16C31 8.8203 25.1797 3 18 3C10.8203 3 5 8.8203 5 16V33H31ZM18 0C9.16344 0 2 7.16344 2 16V35.807C2 35.9136 2.08642 36 2.19303 36H33.807C33.9136 36 34 35.9136 34 35.807V16C34 7.16344 26.8366 0 18 0Z",
                                  },
                                ],
                                "3": [
                                  {
                                    "path": "M8 8C8 8 13.1678 6.70089 14.2197 5.649C15.2716 4.59712 15.2583 2.87836 14.19 1.81004C13.1216 0.741719 11.4029 0.728395 10.351 1.78028C9.29911 2.83216 8 8 8 8Z",
                                  },
                                  {
                                    "path": "M8 8C8 8 13.1678 9.29911 14.2197 10.351C15.2716 11.4029 15.2583 13.1216 14.19 14.19C13.1216 15.2583 11.4029 15.2716 10.351 14.2197C9.29911 13.1678 8 8 8 8Z",
                                  },
                                  {
                                    "path": "M1.78028 10.351C2.83216 9.29911 8 8 8 8C8 8 6.70089 13.1678 5.649 14.2197C4.59712 15.2716 2.87836 15.2583 1.81004 14.19C0.741719 13.1216 0.728395 11.4029 1.78028 10.351Z",
                                  },
                                  {
                                    "path": "M8 8C8 8 2.83216 6.70089 1.78028 5.649C0.728395 4.59712 0.741719 2.87836 1.81004 1.81004C2.87836 0.741719 4.59712 0.728395 5.649 1.78028C6.70089 2.83216 8 8 8 8Z",
                                  },
                                ],
                              }
                            }
                            showBorder={false}
                            size={20}
                          />
                        }
                      >
                        <View
                          transform={
                            [
                              {
                                "scale": 0.5405405405405406,
                              },
                            ]
                          }
                        >
                          <View
                            height={37}
                            width={37}
                            x={0}
                            y={0}
                          >
                            <View
                              colors={
                                [
                                  "#6100FF",
                                  "#76D191",
                                ]
                              }
                              end={null}
                              start={null}
                            />
                          </View>
                          <View
                            color="#D3EBA3"
                            cx={18.5}
                            cy={-13.36111111111111}
                            r={30.833333333333336}
                          >
                            <View
                              blur={15}
                            />
                          </View>
                        </View>
                      </View>
                      <View
                        clip={true}
                        mask={
                          <UniconMask
                            attributeData={
                              {
                                "0": "#6100FF",
                                "1": "#76D191",
                                "2": [
                                  {
                                    "fillType": "evenOdd",
                                    "path": "M31 33V16C31 8.8203 25.1797 3 18 3C10.8203 3 5 8.8203 5 16V33H31ZM18 0C9.16344 0 2 7.16344 2 16V35.807C2 35.9136 2.08642 36 2.19303 36H33.807C33.9136 36 34 35.9136 34 35.807V16C34 7.16344 26.8366 0 18 0Z",
                                  },
                                ],
                                "3": [
                                  {
                                    "path": "M8 8C8 8 13.1678 6.70089 14.2197 5.649C15.2716 4.59712 15.2583 2.87836 14.19 1.81004C13.1216 0.741719 11.4029 0.728395 10.351 1.78028C9.29911 2.83216 8 8 8 8Z",
                                  },
                                  {
                                    "path": "M8 8C8 8 13.1678 9.29911 14.2197 10.351C15.2716 11.4029 15.2583 13.1216 14.19 14.19C13.1216 15.2583 11.4029 15.2716 10.351 14.2197C9.29911 13.1678 8 8 8 8Z",
                                  },
                                  {
                                    "path": "M1.78028 10.351C2.83216 9.29911 8 8 8 8C8 8 6.70089 13.1678 5.649 14.2197C4.59712 15.2716 2.87836 15.2583 1.81004 14.19C0.741719 13.1216 0.728395 11.4029 1.78028 10.351Z",
                                  },
                                  {
                                    "path": "M8 8C8 8 2.83216 6.70089 1.78028 5.649C0.728395 4.59712 0.741719 2.87836 1.81004 1.81004C2.87836 0.741719 4.59712 0.728395 5.649 1.78028C6.70089 2.83216 8 8 8 8Z",
                                  },
                                ],
                              }
                            }
                            overlay={true}
                            showBorder={false}
                            size={20}
                          />
                        }
                      >
                        <View
                          color="#000000"
                          height={20}
                          opacity={0.08}
                          width={20}
                          x={0}
                          y={0}
                        />
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                adjustsFontSizeToFit={true}
                allowFontScaling={true}
                maxFontSizeMultiplier={1.4}
                numberOfLines={1}
                style={
                  {
                    "color": "#222222",
                    "flexShrink": 1,
                    "fontFamily": "Basel-Book",
                    "fontSize": 19,
                    "lineHeight": 24,
                  }
                }
                suppressHighlighting={true}
              >
                0x​82D5...3Fa6
              </Text>
            </View>
          </View>
          <View
            animatedStyle={
              {
                "value": {
                  "opacity": 0,
                },
              }
            }
            cancelable={true}
            collapsable={false}
            disabled={false}
            focusable={true}
            hitSlop={[Function]}
            minPressDuration={0}
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onPress={[Function]}
            onPressIn={[Function]}
            onPressOut={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#CECECE",
                "borderBottomLeftRadius": 999999,
                "borderBottomRightRadius": 999999,
                "borderTopLeftRadius": 999999,
                "borderTopRightRadius": 999999,
                "flexDirection": "column",
                "height": 24,
                "justifyContent": "center",
                "opacity": 0,
                "transform": [
                  {
                    "scale": 1,
                  },
                ],
                "width": 24,
                "zIndex": 1080,
              }
            }
            testID="explore/remove-button"
          >
            <View
              style={
                {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "flexDirection": "column",
                  "height": 2,
                  "width": 10,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
